Fix seeding on BIM edition

aaron-contreras opened the pull request:
Standardizes the Types and Priorities used in the SharedWorkPackageSeeder so seeding is agnostic of specific records present on either the standard or BIM edition of OpenProject.

cbliard committed:
Seed global queries only when there are none

cbliard committed:
Add unit tests to ensure correct record creation

Co-authored-by: Aaron Contreras <aaronlcaq@gmail.com>

aaron-contreras committed:
Reference standardized types and priorities on seeders

Some of the types and priorities used on the shared work package
seeder were not available when seeding the BIM-edition. By using types
and priorities present in both, we fix the not-found reference
error when seeding.

akabiru reviewed (commented):
Hey @aaron-contreras , thanks for looking into this- unfortunately there was still the same `ArgumentError` failure on my end. @wielinde and I looked into a somewhat separate approach in https://github.com/opf/openproject/pull/14807 that sets up fallbacks.

However, I see you added tests that could otherwise be useful, should we harmonise the two PRs?

aaron-contreras commented:
@akabiru I've ran the exact reproduction steps on my machine and there is no `ArgumentError` being raised by doing a `db:seed:replant`. If semantically the Phase type is equivalent to the Epic on the standard edition, I'm all for it but all types and priorities are present on both `standard.yml` and `bim.yml`. Defaulting back to BIM's is okay if so.

I'm curious what you're experiencing as I'm following the exact reproduction steps we found together

akabiru reviewed (approved):
@aaron-contreras I must have failed to checkout properly into this branch, as I've just attempted to reproduce the failure and it's already fixed. 😄 Yeah, let's go ahead with this one. We can link it to https://community.openproject.org/wp/52906

aaron-contreras commented:
We can add a commit to seed with phases instead of epics using the default trick on your PR if that would be the semantically correct type. Let me know :)

akabiru commented:
> We can add a commit to seed with phases instead of epics using the default trick on your PR if that would be the semantically correct type. Let me know :)

Semantically, we discussed this to be more fitting. We also need to back port the fix to 13.3.x so I would update the base branch

---

ref: https://github.com/opf/openproject/pull/14807/commits/46f9f6a3a3b3ca07fc8d424cdccde3a500c6e31a

cbliard committed:
Use better default values for seeded dev data

Standard will use epic and immediate priority for some work packages.
BIM will use phase and high priority instead.

akabiru reviewed (approved):
:shipit:

aaron-contreras commented on app/seeders/source/seed_data.rb:
```diff
@@ -43,15 +43,25 @@ def store_reference(reference, record)
+    elsif default != :__unset__
       default
+    else
+      references = [reference, *fallbacks].map(&:inspect)
+      message = "Nothing registered with #{'reference'.pluralize(references.count)} #{references.to_sentence(locale: false)}"
+      raise ArgumentError, message
```
⭐

cbliard replied:
yeah, maybe a bit overkill but I like nice messages
